### PR TITLE
Add an index setting to limit the maximum number of slices allowed in a scroll request.

### PR DESCRIPTION
**Pull request by jimczi**
This is a follow up for https://github.com/elastic/elasticsearch/pull/18237#discussion_r66096836

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -121,6 +121,12 @@
     public static final Setting<Integer> MAX_REFRESH_LISTENERS_PER_SHARD = Setting.intSetting("index.max_refresh_listeners", 1000, 0,
             Property.Dynamic, Property.IndexScope);
 
+    /**
+     * The maximum number of slices allowed in a scroll request
+     */
+    public static final Setting<Integer> MAX_SLICES_PER_SCROLL = Setting.intSetting("index.max_slices_per_scroll", 10000, 0,
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/IndexSettings.java]: I wonder if we should default it to the number of shards :)

File: core/src/test/java/org/elasticsearch/search/slice/SearchSliceIT.java [hunk 2]
```diff
@@ -172,6 +173,33 @@ public void testInvalidQuery() throws Exception {
             equalTo("`slice` cannot be used outside of a scroll context"));
     }
 
+    public void testMaxSlicesPerScroll() throws Exception {
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/search/slice/SearchSliceIT.java]: can we make this a REST test instead maybe? seems simple though?

File: core/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/IndexSettings.java]: I think we should be very conservative here by default

**Comment by s1monw** [comment]
left some comments! thanks for doing this!

**Comment by jimferenczi** [comment]
@s1monw I pushed another commit to address your comments. The default maximum number of slices is now equal to the number of shards in the requested index.

**Comment by s1monw** [comment]
> @s1monw I pushed another commit to address your comments. The default maximum number of slices is now equal to the number of shards in the requested index.

this might actually not work 😞 if we go through an alias we may have way more slices than shards in a single index. and then we fail half way. So I think we need something different (as you had)? Maybe `1024` ? @jpountz WDTY - sorry for not thinking about it earlier

**Comment by jimferenczi** [comment]
Don't _shrink my scroll ! ;) 
Yep aliases are dangerous. The simple fact that each slice is an independent request is dangerous.
You may also argue that the index could be deleted and re-created between two sliced scroll.
+1 for `1024` and I'll add a note in the docs about aliases and index creation

**Comment by jpountz** [comment]
+1 to a fixed limit, I think it is easier to reason about

**Comment by jimferenczi** [comment]
@jpountz @s1monw I changed the default to 1024.

**Comment by s1monw** [comment]
LGTM

**Pull request merged**